File: system-test/dotnet/SystemTests/Latest/AcceptanceTests/PlaceOrderNegativeIsolatedTest.cs
using SystemTests.Latest.AcceptanceTests.Base;
using Dsl.Core.UseCase;
using Optivem.Testing;
using Xunit;

namespace SystemTests.Latest.AcceptanceTests;

[Collection("Isolated")]
[Trait("Category", "isolated")]
public class PlaceOrderNegativeIsolatedTest : BaseAcceptanceTest
{
    [Theory]
    [Time]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task CannotPlaceOrderWithExpiredCoupon(Channel channel)
    {
        await Scenario(channel)
            .Given().Clock().WithTime("2023-09-01T12:00:00Z")
            .And().Coupon().WithCouponCode("SUMMER2023").WithValidFrom("2023-06-01T00:00:00Z").WithValidTo("2023-08-31T23:59:59Z")
            .When().PlaceOrder().WithCouponCode("SUMMER2023")
            .Then().ShouldFail()
            .ErrorMessage("The request contains one or more validation errors")
            .FieldErrorMessage("couponCode", "Coupon code SUMMER2023 has expired");
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task ShouldRejectOrderPlacedAtYearEnd(Channel channel)
    {
        await Scenario(channel)
            .Given().Clock().WithTime("2026-12-31T23:59:30Z")
            .When().PlaceOrder()
            .Then().ShouldFail()
            .ErrorMessage("Orders cannot be placed between 23:59 and 00:00 on December 31st");
    }
}

File: system-test/java/src/test/java/com/mycompany/myshop/systemtest/latest/acceptance/PlaceOrderNegativeIsolatedTest.java
package com.mycompany.myshop.systemtest.latest.acceptance;

import com.mycompany.myshop.systemtest.latest.acceptance.base.BaseAcceptanceTest;
import com.mycompany.myshop.testkit.channel.ChannelType;
import com.optivem.testing.Channel;
import com.optivem.testing.Isolated;
import org.junit.jupiter.api.TestTemplate;

@Isolated("mutates the shared wall clock to year-end and coupon-expiry times; parallel runs would clash on the clock")
class PlaceOrderNegativeIsolatedTest extends BaseAcceptanceTest {
    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void shouldRejectOrderPlacedAtYearEnd() {
        scenario
                .given().clock()
                    .withTime("2026-12-31T23:59:30Z")
                .when().placeOrder()
                .then().shouldFail()
                    .errorMessage("Orders cannot be placed between 23:59 and 00:00 on December 31st");
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void cannotPlaceOrderWithExpiredCoupon() {
        scenario
                .given().clock()
                    .withTime("2023-09-01T12:00:00Z")
                .and().coupon()
                    .withCouponCode("SUMMER2023")
                    .withValidFrom("2023-06-01T00:00:00Z")
                    .withValidTo("2023-08-31T23:59:59Z")
                .when().placeOrder()
                    .withCouponCode("SUMMER2023")
                .then().shouldFail()
                    .errorMessage("The request contains one or more validation errors")
                    .fieldErrorMessage("couponCode", "Coupon code SUMMER2023 has expired");
    }
}
